### PR TITLE
UI: Add Frontend API function to get value of T-bar

### PR DESCRIPTION
**Pull request by cpyarger**
### Description
This adds a function and a frontend event
function: `obs_frontend_get_tbar_position()`
returns: `int position`
event `OBS_FRONTEND_EVENT_TBAR_VALUE_CHANGED`

### Motivation and Context
When Trying to code for using the tBar it makes more sense to be able to get the position and be notified when it changes, rather then make assumptions

### How Has This Been Tested?
Linux Ubuntu x64 20.10 

### Types of changes

 - Bug fix (non-breaking change which fixes an issue) 
 - New feature (non-breaking change which adds functionality) 

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/api-interface.cpp [hunk 1]
```diff
@@ -165,7 +165,10 @@ struct OBSStudioAPI : obs_frontend_callbacks {
 		QMetaObject::invokeMethod(main, "TBarChanged",
 					  Q_ARG(int, position));
 	}
-
+	int obs_frontend_get_tbar_position(void) override
```
Comment (RytoEX) [inline UI/api-interface.cpp:168]: There should be a blank line between these two functions.

File: UI/obs-frontend-api/obs-frontend-api.cpp [hunk 2]
```diff
@@ -154,6 +154,12 @@ void obs_frontend_set_tbar_position(int position)
 		c->obs_frontend_set_tbar_position(position);
 }
 
+int  obs_frontend_get_tbar_position(void){
```
Comment (RytoEX) [inline UI/obs-frontend-api/obs-frontend-api.cpp:157]: The opening curly brace for a function body should be on its own line.

Comment (RytoEX) [inline UI/obs-frontend-api/obs-frontend-api.cpp:157]: Use one space between the function type and function name.

File: UI/obs-frontend-api/obs-frontend-api.h [hunk 3]
```diff
@@ -49,6 +49,8 @@ enum obs_frontend_event {
 	OBS_FRONTEND_EVENT_RECORDING_UNPAUSED,
 
 	OBS_FRONTEND_EVENT_TRANSITION_DURATION_CHANGED,
+	OBS_FRONTEND_EVENT_TBAR_VALUE_CHANGED,
```
Comment (RytoEX) [inline UI/obs-frontend-api/obs-frontend-api.h:52]: New events are supposed to be added to the end of the enum.  See https://github.com/obsproject/obs-studio/commit/a38cd03e065efbc863bc897012900938aa70beb3.

File: UI/window-basic-main-transitions.cpp [hunk 4]
```diff
@@ -990,8 +991,24 @@ void OBSBasic::TBarChanged(int value)
 
 	obs_transition_set_manual_time(transition,
 				       (float)value / T_BAR_PRECISION_F);
+					   
+
+
```
Comment (RytoEX) [inline UI/window-basic-main-transitions.cpp:996]: Remove unnecessary whitespace.

**Review by RytoEX — CHANGES_REQUESTED** [review]
The appropriate commit prefix for this is "UI: ".  Please use that.

**Comment by cpyarger** [comment]
Sorry about that, requested changes should be complete.

**Comment by cpyarger** [comment]
An example of the t-bar feedback in use

https://www.youtube.com/watch?v=jtC5qPQgAt4

**Review by RytoEX — CHANGES_REQUESTED** [review]
The second commit is fixing an issue in the first commit, so it doesn't need to be separate.  Just squash it into the first commit as a fixup commit.

I'd suggest this for the commit message:
> UI: Add Frontend API function to get value of T-bar
>
> Also add a Frontend API event for when the T-bar's value changes.

That makes it clear what the commit is adding.

**Review by RytoEX — CHANGES_REQUESTED** [review]
The PR looks fine to me visually, though I haven't compiled it and tested it.  It would be nice if you could update the documentation (docs/sphinx/reference-frontend-api.rst) before this gets merged.  That can be a separate commit with the prefix "docs: " or the same commit that notes in the commit message body that the commit also updates documentation.

**Comment by cpyarger** [comment]
@RytoEX Thank you for taking the time to help me with this.

File: docs/sphinx/reference-frontend-api.rst [hunk 5]
```diff
@@ -595,3 +599,27 @@ Functions
...
 
    Reloads the UI canvas and resets libobs video with latest data from profile.
+
+---------------------------------------
+
+.. function:: void obs_frontend_set_tbar_position(int position)
+
+   Sets the value of the transition bar.
+
+   :param position: The position to set the t-bar to
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:609]: Nitpick:  T-bar

File: docs/sphinx/reference-frontend-api.rst [hunk 6]
```diff
@@ -595,3 +599,27 @@ Functions
...
+
+   Release the transition bar.   (like a user releasing their mouse button after moving it).
+
+---------------------------------------
+
+.. function:: int obs_frontend_get_tbar_value(void)
+
+   Gets the value of the transition bar.
+
+   :return: value of the tbar position
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:624]: Nitpick:  T-bar

File: docs/sphinx/reference-frontend-api.rst [hunk 7]
```diff
@@ -595,3 +599,27 @@ Functions
...
+   Sets the value of the transition bar.
+
+   :param position: The position to set the t-bar to
+   :value Range: * 0 - 1023
+
+---------------------------------------
+
+.. function:: void obs_frontend_release_tbar(void);
+
+   Release the transition bar.   (like a user releasing their mouse button after moving it).
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:616]: I'm not a fan of long parenthetical statements, especially with a period after.  Is the example here necessary?  If so, please rephrase it into a regular sentence.

File: docs/sphinx/reference-frontend-api.rst [hunk 8]
```diff
@@ -595,3 +599,27 @@ Functions
...
    Reloads the UI canvas and resets libobs video with latest data from profile.
+
+---------------------------------------
+
+.. function:: void obs_frontend_set_tbar_position(int position)
+
+   Sets the value of the transition bar.
+
+   :param position: The position to set the t-bar to
+   :value Range: * 0 - 1023
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:610]: I think this should be:
```
:param position: The position to set the T-bar to, with a value in 0-1023.
:type position: int
```

The `:value` syntax appears to be meant for ":value: initial value of the variable (text)" rather than possible values.  The `:value` field type seems to be meant for Python rather than C/C++.

https://www.sphinx-doc.org/en/3.x/usage/restructuredtext/domains.html#info-field-lists

File: docs/sphinx/reference-frontend-api.rst [hunk 9]
```diff
@@ -595,3 +599,27 @@ Functions
...
+   Release the transition bar.   (like a user releasing their mouse button after moving it).
+
+---------------------------------------
+
+.. function:: int obs_frontend_get_tbar_value(void)
+
+   Gets the value of the transition bar.
+
+   :return: value of the tbar position
+   :value Range: 0 - 1023
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:625]: Similar to above, I think this should be:
```
:return: The value of the position of the T-bar to, with a value in 0-1023.
:rtype: int
```

The `:value` syntax appears to be meant for ":value: initial value of the variable (text)" rather than possible values.  Also, `:value` and `:return` seem to be more focused on Python than C/C++, but perhaps that a problem for another day.

https://www.sphinx-doc.org/en/3.x/usage/restructuredtext/domains.html#info-field-lists

File: docs/sphinx/reference-frontend-api.rst [hunk 10]
```diff
@@ -595,3 +599,27 @@ Functions
...
+
+.. function:: void obs_frontend_release_tbar(void);
+
+   Release the transition bar.   (like a user releasing their mouse button after moving it).
+
+---------------------------------------
+
+.. function:: int obs_frontend_get_tbar_value(void)
+
+   Gets the value of the transition bar.
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:622]: Nitpick:  "Get" is fine.

File: docs/sphinx/reference-frontend-api.rst [hunk 11]
```diff
@@ -595,3 +599,27 @@ Functions
 .. function:: void obs_frontend_reset_video(void)
 
    Reloads the UI canvas and resets libobs video with latest data from profile.
+
+---------------------------------------
+
+.. function:: void obs_frontend_set_tbar_position(int position)
+
+   Sets the value of the transition bar.
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:607]: Nitpick:  "Set" is fine.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Thank you for adding documentation for the newly added features and also for previously added but undocumented features.  Please see my review notes on the documentation.

File: docs/sphinx/reference-frontend-api.rst (hunk 7 above)
Comment (cpyarger) [reply docs/sphinx/reference-frontend-api.rst:616]: Any suggestions on this one?

Comment (cpyarger) [reply docs/sphinx/reference-frontend-api.rst:616]: maybe 

`Release the T-bar - emulates mouse button release`

Comment (RytoEX) [reply docs/sphinx/reference-frontend-api.rst:616]: I don't use the T-bar, so I had to quickly play with it to figure out what `obs_frontend_release_tbar` does.  It invokes [`OBSBasic::TBarReleased()`](https://github.com/obsproject/obs-studio/blob/e8a1c082daf7b79ec36aa6dfe3a079ba222f3e2b/UI/window-basic-main-transitions.cpp#L922), which appears to determine whether the current transition should be completed (it was dropped near the right), reset (it was dropped near the left), or held (it was dropped outside either of those thresholds), and then if needed, resets the T-bar to 0 and enables/unlocks the transition widgets.  So "releasing their mouse" is correctly describing the process, though it is simplified.

Something like this is a little more descriptive without getting into the weeds, hopefully:
> Emulate a mouse button release on the transition bar and determine transition status.

**Comment by cpyarger** [comment]
@RytoEX I believe this fixes all the documentation changes

**Review by RytoEX — CHANGES_REQUESTED** [review]
I forgot one final nitpick:  use "docs" (lowercase) for the second commit's prefix.  Other than that, visually it looks good to me.  Someone should still compile and test it and make sure the event is emitted when expected and that the Frontend API functions work as expected.

**Comment by cpyarger** [comment]
I believe that should do it :smile_cat:

File: docs/sphinx/reference-frontend-api.rst [hunk 12]
```diff
@@ -595,3 +599,27 @@ Functions
...
+
+---------------------------------------
+
+.. function:: void obs_frontend_release_tbar(void);
+
+   Emulate a mouse button release on the transition bar and determine transition status.
+
+---------------------------------------
+
+.. function:: int obs_frontend_get_tbar_value(void)
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:620]: The function name for this is actually `obs_frontend_get_tbar_position`, not `obs_frontend_get_tbar_value`.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Almost.  A function name in the documentation is incorrect.

File: docs/sphinx/reference-frontend-api.rst (hunk 12 above)
Comment (cpyarger) [reply docs/sphinx/reference-frontend-api.rst]: oops I feel dumb

File: docs/sphinx/reference-frontend-api.rst [hunk 13]
```diff
@@ -595,3 +599,27 @@ Functions
...
+   Emulate a mouse button release on the transition bar and determine transition status.
+
+---------------------------------------
+
+.. function:: int obs_frontend_get_tbar_position(void)
+
+   Get the value of the transition bar.
+
+   :return: The value of the position of the T-bar to, with a value in 0-1023.
+   :rtype: int
```
Comment (RytoEX) [inline docs/sphinx/reference-frontend-api.rst:625]: Nitpick:  These are in the following order in `UI/api-interface.cpp`, `UI/obs-frontend-api/obs-frontend-api.cpp`, and `UI/obs-frontend-api/obs-frontend-api.h`:
```
obs_frontend_release_tbar
obs_frontend_set_tbar_position
obs_frontend_get_tbar_position
```

That should hopefully be my final nitpick.

**Review by RytoEX — CHANGES_REQUESTED** [review]

**Comment by VodBox** [comment]
I am a little worried about returning the t-bar position as an int instead of a float/double between 0.0 and 1.0, especially if TBAR_PRECISION ever gets changed, but that seems incredibly unlikely to happen. Otherwise I have zero issues with this PR.

**Review by VodBox — APPROVED** [review]
